Classes

olgarogova opened the pull request.

olgarogova committed:
Classes

MikhailBelyaev03 commented on src/main/java/Cache.java:
```diff
@@ -0,0 +1,102 @@
+public class Cache<T> {
```
почему все классы без пакетов?

MikhailBelyaev03 commented on .idea/vcs.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
добавь файл .gitignore и укажи файлы и папки, которые не должны попадать в гит.

MikhailBelyaev03 commented on src/main/java/CacheElement.java:
```diff
@@ -0,0 +1,12 @@
+        this.element = element;
+    }
+
+    public T getElement() {
```
где геттеры и сеттеры для всех полей?
еще нужно переопределить equals и hashcode

MikhailBelyaev03 commented on src/main/java/Cache.java:
```diff
@@ -0,0 +1,102 @@
+    public void add(CacheElement<T> element, int index){
+        if (cache[capacity - 1] != null){
+            cache[capacity - 1] = element;
+            if (capacity - 1 - 1 >= 0) System.arraycopy(cache, 1, cache, 0, capacity - 1 - 1);
```
сдвиг массива вынеси в отдельный метод

MikhailBelyaev03 replied:
Всегда используй фигурные скобки, даже если условие/цикл в одну строчку

MikhailBelyaev03 commented on src/main/java/Cache.java:
```diff
@@ -0,0 +1,102 @@
+         Если мы выходим за длину массива, то необходимо удалить самый первый
+         элемент в массиве, сдвинуть весь массив влево и добавить новый элемент в конец массива.
+         */
+    public void add(CacheElement<T> element, int index){
```
Добавлятся должен просто T element

MikhailBelyaev03 commented on src/main/java/Cache.java:
```diff
@@ -0,0 +1,102 @@
+         элемент в массиве, сдвинуть весь массив влево и добавить новый элемент в конец массива.
+         */
+    public void add(CacheElement<T> element, int index){
+        if (cache[capacity - 1] != null){
```
Довольно сложно воспринимается. Лучше пробежаться по всему массиву и проверять свободная ли ячейка в массиве, если свободная то положить туда элемент и сделать return.
а после цикла делаешь сдвиг и вставляешь в последний элемент

MikhailBelyaev03 commented on src/main/java/Cache.java:
```diff
@@ -0,0 +1,102 @@
+    void delete(element) метод удаления элемента из кэша.
+    При удалении мы должны будем сдвинуть оставшуюся часть массива влево.
+     */
+    public void delete(CacheElement<T> element){
```
аналогично в качестве аргумента должен быть про T element

MikhailBelyaev03 commented on src/main/java/Cache.java:
```diff
@@ -0,0 +1,102 @@
+    При удалении мы должны будем сдвинуть оставшуюся часть массива влево.
+     */
+    public void delete(CacheElement<T> element){
+        for (int i = 0; i < capacity; i++) {
```
тут можно воспользоваться методом isPresent, которыый ты уже сделала

MikhailBelyaev03 replied:
и если нашелся элемент, то просто делаешь сдвиг с удалением

MikhailBelyaev03 commented on src/main/java/Cache.java:
```diff
@@ -0,0 +1,102 @@
+    boolean isPresent(element) метод определения есть ли искомый элемент в кэше.
+    */
+    public boolean isPresent(CacheElement<T> element){
+        if (this.capacity == 0){
```
а зачем это условие тут? если capacity==0, то он в цикл не зайдет и вернет false

MikhailBelyaev03 commented on src/main/java/Cache.java:
```diff
@@ -0,0 +1,102 @@
+    /*
+    boolean isPresent(element) метод определения есть ли искомый элемент в кэше.
+    */
+    public boolean isPresent(CacheElement<T> element){
```
аналогично в качестве аргумента должен быть про T element

MikhailBelyaev03 commented on src/main/java/Cache.java:
```diff
@@ -0,0 +1,102 @@
+    При нахождении элемента в кэше его необходимо поместить в конец массива,
+    с учетом сдвига остальных элементов влево.
+    */
+    public CacheElement<T> get(int index){
```
Возвращаем просто T element

MikhailBelyaev03 commented on src/main/java/Cache.java:
```diff
@@ -0,0 +1,102 @@
+    с учетом сдвига остальных элементов влево.
+    */
+    public CacheElement<T> get(int index){
+        for (int i = 0; i < capacity; i++){
```
а зачем ты обнуляешь весь массив??

MikhailBelyaev03 commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,150 @@
+    Если мы достигли предела длины массива, то мы должны увеличить емкость нашего хранилища в 1.5 раза.
+     */
+    public void add(T element) {
+        if (storage[storage.length - 1] == null) {
```
мне кажется лучше сначала проверить на НЕ null и если не null, то увеличивать массив
 и потом без всяких else пробегаешься по полученнгому массиву и вставляешь элемент

MikhailBelyaev03 commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,150 @@
+    После удаляем объект из массива.
+     */
+    public void delete () {
+        if (!checkIfArrayEmpty(storage)) {
```
как то все переусложнено....
просто проверь наличие последнего элемента из storage(для этого можно внести дополнительное поле, которое знает сколько сейчас элементов в storage) в cache с помощью isPresent и если есть удаляй
и потом удаляешь из массива

MikhailBelyaev03 commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,150 @@
+        if (cache.isPresent(index) && cache.get(index) != null) {
+            CacheElement<T> element = cache.get(index);
+            return element.getElement();
+        } else {
```
а если убрать else то поменяется ли логика?)

MikhailBelyaev03 reviewed (commented):
в pom.xml добавь плагин maven-compiler-plugin и в конфигурации плагина укажи версию 11

olgarogova committed:
Changes

olgarogova committed:
Changes - fixed add element to cache, get element from cache and get element from storage

MikhailBelyaev03 commented on .gitignore:
```diff
@@ -0,0 +1,9 @@
+# IntelliJ
+/target/
+
+.idea/compiler.xml
```
можно просто .idea папку указать

MikhailBelyaev03 commented on src/main/java/caching/Cache.java:
```diff
@@ -0,0 +1,106 @@
+         Если мы выходим за длину массива, то необходимо удалить самый первый
+         элемент в массиве, сдвинуть весь массив влево и добавить новый элемент в конец массива.
+         */
+    public void add(T element, int index){
```
У тебя index нигде не используется в методе

MikhailBelyaev03 commented on src/main/java/caching/Cache.java:
```diff
@@ -0,0 +1,106 @@
+        for (int i = 0; i < capacity; i++){
+            if (cache[i] == null) {
+                cache[i] = tCacheElement;
+                elementAdded = true;
```
Тебе не нужен boolean флаг, сделай просто return;

MikhailBelyaev03 commented on src/main/java/caching/Cache.java:
```diff
@@ -0,0 +1,106 @@
+         элемент в массиве, сдвинуть весь массив влево и добавить новый элемент в конец массива.
+         */
+    public void add(T element, int index){
+        CacheElement<T> tCacheElement = new CacheElement<>(element);
```
t - не нужна в имени переменной

MikhailBelyaev03 commented on src/main/java/caching/Cache.java:
```diff
@@ -0,0 +1,106 @@
+        }
+    }
+
+    private void copyArray(CacheElement<T>[] cache){
```
Порядок методов в классе, должен быть сначала паблик, потом протектед и в конце прайват

olgarogova committed:
Cache.java: порядок методов (public, protected, private)
index - поле класса CacheElement, изменены isPresent(index), get(index), add(element, index)

MikhailBelyaev03 commented on src/main/java/caching/Cache.java:
```diff
@@ -0,0 +1,114 @@
+        for (int i = 0; i < capacity; i++){
+            if (cache[i] == null) {
+                cache[i] = cacheElement;
+                cache[i].setIndex(index);
```
Создай у CacheElement новый конструктор, который принимает и element и index)

olgarogova committed:
Added index to CacheElement Constructor
